From be5c755780e52193e00d55cca9475381b3913c35 Mon Sep 17 00:00:00 2001
From: aude <aude.wiki@gmail.com>
Date: Fri, 15 Aug 2014 15:07:14 -0700
Subject: [PATCH] ArrayObject::append() should delegate to
 ArrayObject::offsetSet()

Summary: Match php5 implementation by delegating rather than direct manipulation of internal state.
Fixes issue #3403.
Closes https://github.com/facebook/hhvm/pull/3404

Reviewed By: @fredemmott

Differential Revision: D1486300

Pulled By: svcscm
---
 hphp/system/php/spl/miscellaneous/ArrayObject.php |  2 +-
 hphp/test/slow/array_object/append.php            | 15 +++++++++++++++
 hphp/test/slow/array_object/append.php.expect     | 18 ++++++++++++++++++
 3 files changed, 34 insertions(+), 1 deletion(-)

diff --git a/hphp/system/php/spl/miscellaneous/ArrayObject.php b/hphp/system/php/spl/miscellaneous/ArrayObject.php
index d405cc0..ed020c1 100644
--- a/hphp/system/php/spl/miscellaneous/ArrayObject.php
+++ b/hphp/system/php/spl/miscellaneous/ArrayObject.php
@@ -72,7 +72,7 @@ class ArrayObject implements IteratorAggregate, ArrayAccess,
         'use ArrayObject::offsetSet() instead'
       );
     }
-    $this->storage[] = $value;
+    $this->offsetSet(null, $value);
   }
 
   // This doc comment block generated by idl/sysdoc.php
diff --git a/hphp/test/slow/array_object/append.php b/hphp/test/slow/array_object/append.php
index da7f7ca..ba5c292 100644
--- a/hphp/test/slow/array_object/append.php
+++ b/hphp/test/slow/array_object/append.php
@@ -1,6 +1,21 @@
 <?php
 
+class ExtendedArrayObject extends ArrayObject {
+  public function offsetSet($key, $value) {
+    $key = 'q1';
+    parent::offsetSet($key, $value);
+  }
+}
+
 $arrayobj = new ArrayObject(array('first','second','third'));
 $arrayobj->append('fourth');
 $arrayobj->append(array('five', 'six'));
 var_dump($arrayobj);
+
+$arrayobj = new ExtendedArrayObject(array('y'));
+$arrayobj->append('x');
+var_dump($arrayobj);
+
+$arrayobj = new ExtendedArrayObject(array('q2' => 'y'));
+$arrayobj->append('z');
+var_dump($arrayobj);
diff --git a/hphp/test/slow/array_object/append.php.expect b/hphp/test/slow/array_object/append.php.expect
index 194af66..45e2fdb 100644
--- a/hphp/test/slow/array_object/append.php.expect
+++ b/hphp/test/slow/array_object/append.php.expect
@@ -18,3 +18,21 @@ object(ArrayObject)#1 (1) {
     }
   }
 }
+object(ExtendedArrayObject)#2 (1) {
+  ["storage":"ArrayObject":private]=>
+  array(2) {
+    [0]=>
+    string(1) "y"
+    ["q1"]=>
+    string(1) "x"
+  }
+}
+object(ExtendedArrayObject)#3 (1) {
+  ["storage":"ArrayObject":private]=>
+  array(2) {
+    ["q2"]=>
+    string(1) "y"
+    ["q1"]=>
+    string(1) "z"
+  }
+}
-- 
2.1.0.rc1

